ethers-solc/src/resolver.rs

```diff
@@ -0,0 +1,518 @@
+                                let node = read_node(&target)?;
+                                unresolved.push_back((target.clone(), node));
+                                let idx = index.len();
+                                index.insert(target.clone(), idx);
+                                resolved_imports.push(idx);
+                            }
+                        }
+                        Err(err) => {
+                            tracing::trace!("failed to resolve relative import \"{:?}\"", err);
+                        }
+                    }
+                } else {
+                    // resolve library file
+                    if let Some(lib) = paths.resolve_library_import(import.as_ref()) {
+                        if let Some(idx) = index.get(&lib).cloned() {
+                            resolved_imports.push(idx);
+                        } else {
+                            // imported file is not part of the input files
+                            let node = read_node(&lib)?;
+                            unresolved.push_back((lib.clone(), node));
+                            let idx = index.len();
+                            index.insert(lib.clone(), idx);
+                            resolved_imports.push(idx);
+                        }
+                    } else {
+                        tracing::trace!(
+                            "failed to resolve library import \"{:?}\"",
+                            import.display()
+                        );
+                    }
+                }
+            }
+            nodes.push(node);
+            edges.push(resolved_imports);
+        }
+
+        Ok(Graph { nodes, edges, indices: index, num_input_files, root: paths.root.clone() })
+    }
+
+    /// Resolves the dependencies of a project's source contracts
+    pub fn resolve(paths: &ProjectPathsConfig) -> Result<Graph> {
+        Self::resolve_sources(paths, paths.read_input_files()?)
+    }
+}
+
+#[cfg(all(feature = "svm", feature = "async"))]
+impl Graph {
+    /// Returns all input files together with their appropriate version.
+    ///
+    /// First we determine the compatible version for each input file (sources and test, see
+    /// `Self::resolve`) Then we also add any other dependencies
+    pub fn into_sources_by_version(self, offline: bool) -> Result<VersionedSources> {
+        /// insert the imports of the given node into the sources map
+        /// There can be following graph:
+        /// `A(<=0.8.10) imports C(>0.4.0)` and `B(0.8.11) imports C(>0.4.0)`
+        /// where `C` is a library import, in which case we assign `C` only to the first input file.
+        /// However, it's not required to include them in the solc `CompilerInput` as they would get
+        /// picked up by solc otherwise, but we add them so we can create a corresponding
+        /// cache entry for them as well. This can be optimized however
```

Comment:
We're basically double compiling some files?

Reply:
I think that's possible in scenarios like:
`A(<0.8.10) imports C (>0.4.0)`
`B(=0.8.11) imports C (>0.4.0)`

where A and B are input contracts and C is lib import.
In which case we can't compile A and B together but both require C.
But we only have a single instance of C in our dependency graph.
So I think it's fine to only include it in 1 `Solc -> Sources` set, like `0.8.11 -> (B,C)`, because solc will resolve C for A itself if not provided in the standard json input.

the reason why I decided against cloning here is, that these `Sources` can cause some overhead when we process the cache entries.
I wanted to optimize that first.